[WIP] Quantized Average Pool kernel

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23143 [WIP] Quantized Average Pool kernel**

Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

Commit:
[WIP] Quantized Average Pool kernel

zafartahirov commented:
This is a draft. Most of the stuff is copied from the `AdaptiveAvgPool.cpp`, but code reuse was hard because of slightly different semantics in qtypes.

TODO:
- Unittests
- Functionals
- Modules

Commit:
Update on "[WIP] Quantized Average Pool kernel"

[WIP] Quantized Average Pool kernel

gh-metadata: pytorch pytorch 23143 gh/zafartahirov/12/head

Commit:
Update on "[WIP] Quantized Average Pool kernel"

[WIP] Quantized Average Pool kernel

gh-metadata: pytorch pytorch 23143 gh/zafartahirov/12/head

Commit:
Update on "[WIP] Quantized Average Pool kernel"

[WIP] Quantized Average Pool kernel

gh-metadata: pytorch pytorch 23143 gh/zafartahirov/12/head

Test:

`python test/test_quantized.py TestQuantizedOps -v`

```
test_adaptive_avg_pool2d (__main__.TestQuantizedOps) ... ../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
../aten/src/ATen/native/AdaptiveAveragePooling.cpp:337: UserWarning: Quantized mean is WIP!
  TORCH_WARN("Quantized mean is WIP! Falling back to AdaptiveAvgPool");
ok
test_cat (__main__.TestQuantizedOps) ... ok
test_max_pool2d (__main__.TestQuantizedOps) ... ok
test_qadd_relu_different_qparams (__main__.TestQuantizedOps) ... ok
test_qadd_relu_same_qparams (__main__.TestQuantizedOps) ... ok
test_qrelu (__main__.TestQuantizedOps) ... ok

----------------------------------------------------------------------
Ran 6 tests in 1.951s

OK
```

zafartahirov commented:
Added a warning for the `mean` computation

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:
```diff
@@ -0,0 +1,222 @@
+          }
+
+          /* set output to local average */
+          *op = scalar_t(clamp_and_cast<underlying_t>(sum / kW / kH));
```
if you're just doing avg - there's no need to clamp, right?

dzhulgakov commented on aten/src/ATen/native/AdaptiveAveragePooling.cpp:
```diff
@@ -312,7 +312,12 @@ namespace {
-    auto output = at::empty({0}, input.options());
+    Tensor output;
+    if (input.is_quantized()) {
+      output = at::_empty_affine_quantized({0}, input.options());
```
don't  you need to set quantizer here?

jerryzh168 replied:
I believe empty_affine_quantized takes scale and zero_point as well

dzhulgakov reviewed (changes requested):
I wonder if it's possible to reduce amount of copy-pasta. If I understand correctly, you don't really need to do clamping. If so, we could just add "accumulate_t" argument to the original function and be happy with it

zafartahirov commented:
> I wonder if it's possible to reduce amount of copy-pasta. If I understand correctly, you don't really need to do clamping. If so, we could just add "accumulate_t" argument to the original function and be happy with it

I tried putting `QuantizedCPU` support in the original `AdaptiveAvgPool` file (spent the whole Saturday trrying) -- failed miserably.

jerryzh168 commented:
let's land the current code first and refactor later.

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:
```diff
@@ -0,0 +1,222 @@
+          }
+
+          /* set output to local average */
+          *op = scalar_t(clamp_and_cast<underlying_t>(sum / kW / kH));
```
Where is the output quantized back to 8 bits?

zafartahirov replied:
Why do we need to quantize it back?

zafartahirov commented:
> I wonder if it's possible to reduce amount of copy-pasta. If I understand correctly, you don't really need to do clamping. If so, we could just add "accumulate_t" argument to the original function and be happy with it

I created a task for myself to refactor it to avoid as much copy/paste stuff as possible.

Commit:
Update on "[WIP] Quantized Average Pool kernel"

[WIP] Quantized Average Pool kernel

gh-metadata: pytorch pytorch 23143 gh/zafartahirov/12/head

Test:

`python test/test_quantized.py TestQuantizedOps -v`

```
test_adaptive_avg_pool2d (__main__.TestQuantizedOps) ... ok
test_cat (__main__.TestQuantizedOps) ... ok
test_max_pool2d (__main__.TestQuantizedOps) ... ok
test_qadd_relu_different_qparams (__main__.TestQuantizedOps) ... ok
test_qadd_relu_same_qparams (__main__.TestQuantizedOps) ... ok
test_qrelu (__main__.TestQuantizedOps) ... ok

----------------------------------------------------------------------
Ran 6 tests in 1.857s

OK
```

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

dzhulgakov commented on aten/src/ATen/native/AdaptiveAveragePooling.cpp:
```diff
@@ -324,12 +329,18 @@ namespace {
+      // dimensions, which can be done more efficiently
+      int64_t mean_size = input.size(-1) * input.size(-2);
+      Tensor out = input.contiguous().view({-1, mean_size});
+      if (out.is_quantized()) {
```
move it to the if statement (checking input.is_quantized) - you'd skip allocating the output

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:
```diff
@@ -0,0 +1,199 @@
+          }
+
+          /* set output to local average */
+          *op = scalar_t(static_cast<underlying_t>(sum / kW / kH));
```
I'm not an expert - but should it do some kind of rounding?

zafartahirov replied:
I was thinking the casting should take care of it. But I might be wrong -- casting truncates (floor). Maybe I do need a round -- I'll check

zafartahirov replied:
You were right -- without proper rounding, there was a corner case with 1-off error!

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:
```diff
@@ -0,0 +1,202 @@
+    output_sizes = {sizeB, sizeD, osizeH, osizeW};
+  }
+  if (!output.is_quantized() || output.sizes() != output_sizes) {
+    output = at::_empty_affine_quantized(output_sizes, input.options(),
```
hm, is it semantics we want? if the output the wrong size - inherit quantizer from input, otherwise from output? it's not how majority of ops work in PyTorch and I don't think it's a good behavior.

probably a better approach is to force output to be already quantized and error out if the size is wrong. @gchanan

zafartahirov replied:
So basically throw an error at this point?

jerryzh168 replied:
yeah, if output is provided, then it must be quantized and has the correct size.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:
```diff
@@ -0,0 +1,202 @@
+};
+
+static auto registry = torch::RegisterOperators().op(
+  "quantized::adaptive_avg_pool2d(Tensor qx, int[] output_size) -> Tensor",
```
since you already use native_functions, I don't think you need to register it again

zafartahirov replied:
I agree -- we were "double-registering" all the native_functions. I wanted to clean them up at some point :)

Commit:
Update on "[WIP] Quantized Average Pool kernel"


Differential Revision: [D16406281](https://our.internmc.facebook.com/intern/diff/D16406281)

dzhulgakov reviewed (approved):
Copypasta of implementation is still a bit annoying, but should be good to go

zafartahirov commented:
> Copypasta of implementation is still a bit annoying, but should be good to go

We will create a bootcamp task for it 😃
